feat(elasticsearch): support @elastic/elasticsearch client

trentm opened the pull request:
Based on earlier work in:
https://github.com/elastic/apm-agent-nodejs/pull/1266


Fixes #968

apmmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-1877/detail/PR-1877/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-1877/detail/PR-1877/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-1877/detail/PR-1877/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-1877/detail/PR-1877/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://apm-agent-nodejs_1877.docs-preview.app.elstc.co/diff)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats

* Build Cause: `[Pull request #1877 opened]`

* Start Time: 2020-11-17T16:25:20.757+0000

* Duration: 12 min 52 sec




#### Test stats :test_tube:

| Test         | Results                         |
| ------------ | :-----------------------------: |
| Failed       | 0  |
| Passed       | 16398  |
| Skipped      | 0 |
| Total        | 16398   |


</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY-->


### Steps errors [![56](https://img.shields.io/badge/56%20-red)](https://apm-ci.elastic.co/blue/organizations/jenkins/apm-agent-nodejs%2Fapm-agent-nodejs-mbp%2FPR-1877/detail/PR-1877/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>

  > Show only the first 10 steps failures
  
  
  
  
#### `Run Tests`
<ul>
<li>Took 2 min 21 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-1877/runs/1/steps/188/log/?start=0">here</a></li>

</ul>
  
  
  
  
#### `Run Tests`
<ul>
<li>Took 0 min 26 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-1877/runs/1/steps/377/log/?start=0">here</a></li>

</ul>
  
  
  
  
#### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-1877/runs/1/steps/424/log/?start=0">here</a></li>
<li>Description: <code>hudson.AbortException: script returned exit code 1</code></l1>
</ul>
  
  
  
  
#### `Run Tests`
<ul>
<li>Took 2 min 15 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-1877/runs/1/steps/193/log/?start=0">here</a></li>

</ul>
  
  
  
  
#### `Run Tests`
<ul>
<li>Took 0 min 24 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-1877/runs/1/steps/374/log/?start=0">here</a></li>

</ul>
  
  
  
  
#### `Error signal`
<ul>
<li>Took 0 min 1 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-1877/runs/1/steps/412/log/?start=0">here</a></li>
<li>Description: <code>hudson.AbortException: script returned exit code 1</code></l1>
</ul>
  
  
  
  
#### `Run Tests`
<ul>
<li>Took 3 min 35 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-1877/runs/1/steps/234/log/?start=0">here</a></li>

</ul>
  
  
  
  
#### `Run Tests`
<ul>
<li>Took 1 min 31 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-1877/runs/1/steps/394/log/?start=0">here</a></li>

</ul>
  
  
  
  
#### `Error signal`
<ul>
<li>Took 0 min 0 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-1877/runs/1/steps/555/log/?start=0">here</a></li>
<li>Description: <code>hudson.AbortException: script returned exit code 1</code></l1>
</ul>
  
  
  
  
#### `Run Tests`
<ul>
<li>Took 3 min 32 sec . View more details on <a href="https://apm-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/pipelines/apm-agent-nodejs-mbp/pipelines/PR-1877/runs/1/steps/246/log/?start=0">here</a></li>

</ul>
  
  </p>
  </details>



<!-- LOG OUTPUT-->
### Log output

<details><summary>Expand to view the last 100 lines of log output</summary>
<p>

```
[2020-11-17T16:36:48.614Z]     Yaml found at: .codecov.yml
[2020-11-17T16:36:48.614Z] ==> Running gcov in . (disable via -X gcov)
[2020-11-17T16:36:49.139Z] ==> Python coveragepy not found
[2020-11-17T16:36:49.139Z] ==> Searching for coverage reports in:
[2020-11-17T16:36:49.139Z]     + .
[2020-11-17T16:36:49.139Z] --> No coverage report found.
[2020-11-17T16:36:49.139Z]     Please visit http://docs.codecov.io/docs/supported-languages
[2020-11-17T16:36:50.428Z] + curl -sSLo codecov.sh https://codecov.io/bash
[2020-11-17T16:36:52.127Z] + curl -sSLo codecov.sh https://codecov.io/bash
[2020-11-17T16:36:52.179Z] [INFO] Codecov: Sending data...
[2020-11-17T16:36:53.573Z] + bash codecov.sh
[2020-11-17T16:36:53.573Z] 
[2020-11-17T16:36:53.573Z]   _____          _
[2020-11-17T16:36:53.573Z]  / ____|        | |
[2020-11-17T16:36:53.573Z] | |     ___   __| | ___  ___ _____   __
[2020-11-17T16:36:53.573Z] | |    / _ \ / _` |/ _ \/ __/ _ \ \ / /
[2020-11-17T16:36:53.573Z] | |___| (_) | (_| |  __/ (_| (_) \ V /
[2020-11-17T16:36:53.573Z]  \_____\___/ \__,_|\___|\___\___/ \_/
[2020-11-17T16:36:53.573Z]                               Bash-20201116-eea08d9
[2020-11-17T16:36:53.573Z] 
[2020-11-17T16:36:53.573Z] 
[2020-11-17T16:36:53.573Z] ==> Jenkins CI detected.
[2020-11-17T16:36:53.573Z]     project root: .
[2020-11-17T16:36:53.573Z] --> token set from env
[2020-11-17T16:36:53.573Z]     Yaml found at: .codecov.yml
[2020-11-17T16:36:53.574Z] ==> Running gcov in . (disable via -X gcov)
[2020-11-17T16:36:53.574Z] ==> Python coveragepy not found
[2020-11-17T16:36:53.574Z] ==> Searching for coverage reports in:
[2020-11-17T16:36:53.574Z]     + .
[2020-11-17T16:36:53.574Z] --> No coverage report found.
[2020-11-17T16:36:53.574Z]     Please visit http://docs.codecov.io/docs/supported-languages
[2020-11-17T16:36:53.602Z] Failed in branch Node.js-12.0-async-hooks-false
[2020-11-17T16:36:54.441Z] + bash codecov.sh
[2020-11-17T16:36:54.441Z] 
[2020-11-17T16:36:54.441Z]   _____          _
[2020-11-17T16:36:54.441Z]  / ____|        | |
[2020-11-17T16:36:54.441Z] | |     ___   __| | ___  ___ _____   __
[2020-11-17T16:36:54.441Z] | |    / _ \ / _` |/ _ \/ __/ _ \ \ / /
[2020-11-17T16:36:54.441Z] | |___| (_) | (_| |  __/ (_| (_) \ V /
[2020-11-17T16:36:54.441Z]  \_____\___/ \__,_|\___|\___\___/ \_/
[2020-11-17T16:36:54.441Z]                               Bash-20201116-eea08d9
[2020-11-17T16:36:54.441Z] 
[2020-11-17T16:36:54.441Z] 
[2020-11-17T16:36:54.441Z] ==> Jenkins CI detected.
[2020-11-17T16:36:54.442Z]     project root: .
[2020-11-17T16:36:54.442Z] --> token set from env
[2020-11-17T16:36:54.442Z]     Yaml found at: .codecov.yml
[2020-11-17T16:36:54.442Z] ==> Running gcov in . (disable via -X gcov)
[2020-11-17T16:36:54.548Z] ==> Python coveragepy not found
[2020-11-17T16:36:54.548Z] ==> Searching for coverage reports in:
[2020-11-17T16:36:54.548Z]     + .
[2020-11-17T16:36:54.548Z] --> No coverage report found.
[2020-11-17T16:36:54.548Z]     Please visit http://docs.codecov.io/docs/supported-languages
[2020-11-17T16:36:56.516Z] + curl -sSLo codecov.sh https://codecov.io/bash
[2020-11-17T16:36:57.400Z] Failed in branch Node.js-8.6-async-hooks-false
[2020-11-17T16:36:58.189Z] + bash codecov.sh
[2020-11-17T16:36:58.189Z] 
[2020-11-17T16:36:58.189Z]   _____          _
[2020-11-17T16:36:58.189Z]  / ____|        | |
[2020-11-17T16:36:58.190Z] | |     ___   __| | ___  ___ _____   __
[2020-11-17T16:36:58.190Z] | |    / _ \ / _` |/ _ \/ __/ _ \ \ / /
[2020-11-17T16:36:58.190Z] | |___| (_) | (_| |  __/ (_| (_) \ V /
[2020-11-17T16:36:58.190Z]  \_____\___/ \__,_|\___|\___\___/ \_/
[2020-11-17T16:36:58.190Z]                               Bash-20201116-eea08d9
[2020-11-17T16:36:58.190Z] 
[2020-11-17T16:36:58.190Z] 
[2020-11-17T16:36:58.190Z] ==> Jenkins CI detected.
[2020-11-17T16:36:58.190Z]     project root: .
[2020-11-17T16:36:58.190Z] --> token set from env
[2020-11-17T16:36:58.190Z]     Yaml found at: .codecov.yml
[2020-11-17T16:36:58.190Z] ==> Running gcov in . (disable via -X gcov)
[2020-11-17T16:36:58.190Z] ==> Python coveragepy not found
[2020-11-17T16:36:58.190Z] ==> Searching for coverage reports in:
[2020-11-17T16:36:58.190Z]     + .
[2020-11-17T16:36:58.190Z] --> No coverage report found.
[2020-11-17T16:36:58.190Z]     Please visit http://docs.codecov.io/docs/supported-languages
[2020-11-17T16:36:59.083Z] Failed in branch Node.js-8
[2020-11-17T16:36:59.467Z] Failed in branch Node.js-8.6
[2020-11-17T16:37:00.180Z] Failed in branch Node.js-8-async-hooks-false
[2020-11-17T16:37:04.733Z] Stage "TAV Test" skipped due to earlier failure(s)
[2020-11-17T16:37:05.130Z] Stage "Edge Test" skipped due to earlier failure(s)
[2020-11-17T16:37:05.640Z] Stage "Nightly Test" skipped due to earlier failure(s)
[2020-11-17T16:37:05.721Z] Stage "Nightly Test - No async hooks" skipped due to earlier failure(s)
[2020-11-17T16:37:05.841Z] Stage "RC Test" skipped due to earlier failure(s)
[2020-11-17T16:37:05.925Z] Stage "RC Test - No async hooks" skipped due to earlier failure(s)
[2020-11-17T16:37:06.743Z] Failed in branch Nightly Test
[2020-11-17T16:37:06.816Z] Failed in branch Nightly Test - No async hooks
[2020-11-17T16:37:06.901Z] Failed in branch RC Test
[2020-11-17T16:37:07.034Z] Failed in branch RC Test - No async hooks
[2020-11-17T16:37:08.106Z] Stage "Integration Tests" skipped due to earlier failure(s)
[2020-11-17T16:37:08.491Z] Stage "Release" skipped due to earlier failure(s)
[2020-11-17T16:37:08.613Z] Stage "Release" skipped due to earlier failure(s)
[2020-11-17T16:37:09.413Z] Stage "Benchmarks" skipped due to earlier failure(s)
[2020-11-17T16:37:10.579Z] Running on Jenkins in /var/lib/jenkins/workspace/ejs_apm-agent-nodejs-mbp_PR-1877
[2020-11-17T16:37:10.827Z] [INFO] getVaultSecret: Getting secrets
[2020-11-17T16:37:11.347Z] Masking supported pattern matches of $VAULT_ADDR or $VAULT_ROLE_ID or $VAULT_SECRET_ID
[2020-11-17T16:37:13.477Z] + chmod 755 generate-build-data.sh
[2020-11-17T16:37:13.477Z] + ./generate-build-data.sh https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-1877/ https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-1877/runs/1 FAILURE 712387
[2020-11-17T16:37:14.058Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-1877/runs/1/steps/?limit=10000 -o steps-info.json
[2020-11-17T16:37:17.292Z] INFO: curl https://apm-ci.elastic.co/blue/rest/organizations/jenkins/pipelines/apm-agent-nodejs/apm-agent-nodejs-mbp/PR-1877/runs/1/tests/?status=FAILED -o tests-errors.json

```

</p>
</details>



<!--COMMENT_GENERATED_WITH_ID_comment.id-->

trentm commented:
This is ready to review now. There are a few "TODO" comments in there:

```
% git diff master.. | grep TODO
+// TODO: ask Tomas about apiName addition to RequestEvent
+        // TODO set "span.outcome" (from err and result.statusCode)
+          // TODO: rebuild error object to avoid serialization issues?
+    // TODO test that httpSpan is child of esSpan? currently it isn't
```

that I don't think *all* need to be done now.

@delvedor You are welcome to review and/or play with this, but I don't expect you to.

trentm commented:
review note: the linting "error" will go away when these commits are squashed

astorm commented on lib/instrumentation/modules/@elastic/elasticsearch.js:
```diff
@@ -0,0 +1,124 @@
+const { getDBDestination } = require('../../context')
+
+// TODO: ask Tomas about apiName addition to RequestEvent
+const pathIsAQuery = /\/_((search|msearch)(\/template)?|count)$/
```
Calling out some offline discussions here -- this test means that we'll only capture the full request bodies for REST queries whose paths match the following _exactly_. 

```
_search
_msearch
_search/template
_msearch/template
_count
```

We may want to expand this after discussions with stakeholders.

trentm replied:
I've also asked @delvedor about perhaps changing elasticsearch-js to provide a field in the 'request' event that is better than just regex matching the path (which can have user data in it).

I've also got this change coming:

```
// Allow the following URL path components:
// - _search
// - _msearch
// - _search/template
// - _msearch/template
// - _count
// - _async_search
const pathIsAQuery = /\/(_search|_msearch|_count|_async_search)(\/|$)/
```

to add `_async_search` because Kibana is using that.

astorm reviewed (commented):
I took an initial pass at this -- it looks like a solid start and captures the functionality of the  previous instrumentation and @delvedor's first pass at this well.  The use of the `WeakMap` is a nice call-out as well -- Trust your ~Mechanic~ Garbage Collector :) 

It doesn't look like sub-classing is the common approach for instrumentation in the agent -- normally we'd identify the request method, wrap it, start our business, call the original method, and then clean up our business.  

I think the sub-classing/event approach can work though -- and in some ways seems less intrusive than the direct monkey-patching we normally do. There's a few additional defensive measures we'll want to take, but otherwise I'm down with this approach.

trentm commented on lib/instrumentation/modules/@elastic/elasticsearch.js:
```diff
@@ -0,0 +1,124 @@
+  class ApmClient extends elasticsearch.Client {
+    constructor (opts) {
+      super(opts)
+
```
Heh, I'd had a comment for that myself, then removed it. I doubt this signature will change, but it doesn't hurt, so sounds good.

trentm commented on lib/instrumentation/modules/@elastic/elasticsearch.js:
```diff
@@ -0,0 +1,124 @@
+    return span
+  }
+
+  class ApmClient extends elasticsearch.Client {
```
Thanks. I'll add a guard for that to fail gracefully:

```
  if (!elasticsearch.Client) {
    agent.logger.debug('@elastic/elasticsearch@%s is not supported (no `elasticsearch.Client`) - aborting...', version)
    return elasticsearch
  }
```

In testing it looks like this:

```
$ ELASTIC_APM_LOG_LEVEL=trace node my-test-app.js
...
shimming https@10.23.0 module
shimming https.Server.prototype.emit function
shimming https.request function
shimming https.get function
shimming @elastic/elasticsearch@7.10.0 module
@elastic/elasticsearch@7.10.0 is not supported (no `elasticsearch.Client`) - aborting...   <---- HERE
shimming elasticsearch@16.7.1 module
shimming elasticsearch.Transport.prototype.request
...
```

trentm commented:
> I think the sub-classing/event approach can work though -- and in some ways seems less intrusive than the direct monkey-patching we normally do.

My understanding is this was intentional. The 'request' and 'response' events described at https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/observability.html were, at least in part, intended for this kind of thing.

trentm commented:
Remaining TODOs before merging:

- [ ] Look into getting the http span(s) to be a *child* of the Elasticsearch span
- [ ] Decide on error serialization. From the [original PR](https://github.com/elastic/apm-agent-nodejs/pull/1266#issue-305457143): "we should figure out how to serialize the errors to avoid the loss of important information"


Related TODOs that can be done separate from this PR (IMHO):

- [ ] Set "span.outcome" (from err and result.statusCode). I'll take a stab at this, but might defer it to https://github.com/elastic/apm-agent-nodejs/issues/1814
- [ ] Trent to possibly open a ticket/PR on elasticsearch-js to have an `apiName` field on the "request" event data to identify what Elasticsearch API endpoint is being called. This would make the current `pathIsAQuery` filtering (to decide if `context.db` is added to an APM span) more efficient and precise.
- [ ] Trent to open a separate ticket/PR to look at adding a "prepare-request" (or whatever name) event from elasticsearch-js to include request setup time. This was explored in the original PR: https://github.com/elastic/apm-agent-nodejs/pull/1266/files#diff-8b8cdefabed92970b05a687c766e3c51678ea17686b219d93ff63043f40a6538R27 and ... I can't find the "elasticsearch-js/update-events" branch anymore (perhaps since deleted).

trentm commented:
I am moving the TODO item `Look into getting the http span(s) to be a child of the Elasticsearch span` to a *separate* ticket, because handling that is a bigger change to agent behaviour: https://github.com/elastic/apm-agent-nodejs/issues/1889

delvedor commented on lib/instrumentation/modules/@elastic/elasticsearch.js:
```diff
@@ -0,0 +1,140 @@
+// - _msearch/template
+// - _count
+// - _async_search
+const pathIsAQuery = /\/(_search|_msearch|_count|_async_search)(\/|$)/
```
The `POST /_sql` path can contain a query as well. ([documentation](https://www.elastic.co/guide/en/elasticsearch/client/javascript-api/current/api-reference.html#_sql_query))
```suggestion
const pathIsAQuery = /\/(_search|_msearch|_count|_async_search|_sql)(\/|$)/
```

delvedor reviewed (commented):
Great work! I love the use of the WeakMap.

Regarding the `preRequest` event I'll work on it soon. I'm wondering if it makes more sense to have a `serialization` and `deserialization` events tho.
So the order would be:
```
serialization
  │ (serialization and compression happens between those two events)
  └─▶ request
	  │ (actual time spent over the wire)
	  └─▶ deserialization
		  │ (deserialization and decompression happens between those two events)
		  └─▶ response
```
What do you think?

delvedor commented:
I've opened https://github.com/elastic/elasticsearch-js/pull/1364, which adds the api name to the request metadata so you can avoid using the regex and get better performances out of the box :)
If you are ok with the addition, it will land in 7.11.

trentm commented:
> I'm wondering if it makes more sense to have a `serialization` and `deserialization` events tho.

From our discussion earlier, this sounds great. Thanks for working on this.

vigneshshanmugam commented on lib/instrumentation/modules/@elastic/elasticsearch.js:
```diff
@@ -0,0 +1,117 @@
+            }
+          }
+
+          delete err.data // possibly large or private user data
```
might trigger a deopt, cant we just discard it or ignore specific fields from getting captured as part of error instead of deleting them and triggering deopt.

trentm replied:
@vigneshshanmugam Thanks for bringing it up. I haven't played with tracing deopts at all. Currently the node.js agent's `captureError` doesn't provide a way to guide which error fields should be captured. I can look into adding that capability.

trentm replied:
https://github.com/elastic/apm-agent-nodejs/pull/1912 is my hoped-for solution for this.

trentm replied:
That is merged now and this PR moved to using it: https://github.com/elastic/apm-agent-nodejs/pull/1877/commits/20984f22cc073a08d85a551a1b8ddbdf0b312bfd

vigneshshanmugam reviewed (commented):
Great work and love the use of WeakMap.

astorm reviewed (approved):
👍 I see my comments from the previous review have been addressed.  Also, Re:

> I am moving the TODO item Look into getting the http span(s) to be a child of the Elasticsearch span to a separate ticket, because handling that is a bigger change to agent behaviour: #1889

This feels like the correct approach, and that it would be better to get this PR landed and then address the span parenting in a separate issue.

trentm commented:
jenkins run the module tests for elasticsearch,@elastic/elasticsearch

trentm commented:
jenkins run the module tests for elasticsearch,@elastic/elasticsearch

trentm commented:
jenkins run the module tests for elasticsearch,@elastic/elasticsearch

trentm commented:
jenkins run the module tests for ALL

trentm commented:
Since the last time y'all reviewed:

- I moved to using `apm.captureError(..., {captureAttributes: false})` to avoid deopt (per @vigneshshanmugam comment): https://github.com/elastic/apm-agent-nodejs/pull/1877/files#diff-8b8cdefabed92970b05a687c766e3c51678ea17686b219d93ff63043f40a6538R92-R97
- I fixed a side issue where `error.exception.module` was not handling "@ns/package" namespaced packages: https://github.com/elastic/apm-agent-nodejs/pull/1877/files#diff-9ceb00d437ddb7c1de1041eea52561f5e7166af4375d35d769590caace91ed31R282-R307
- I removed races from the added tests.

This is ready for re-review when you have the time. Thanks.

astorm commented on lib/instrumentation/modules/elasticsearch.js:
```diff
@@ -2,8 +2,7 @@
 var { getDBDestination } = require('../context')
-
-var queryRegexp = /_((search|msearch)(\/template)?|count)$/
+const { setElasticsearchDbContext } = require('../elasticsearch-shared')
```
👍 to pulling this into a shared lib across both versions of the client.

astorm commented on lib/parsers.js:
```diff
@@ -279,13 +279,32 @@ function getCulprit (frames) {
+    // Normalize the module name separator to '/', even on Windows.
+    moduleName += '/' + match[3]
+  }
+  return moduleName
```
I see here we're modifying `getModule` so it works with `@namespaced/module-names` as well as with windows based filesystems. I presume this is to capture `@elastic/elasticsearch` things.

Do you think this changes requires any additional test fixtures in `test/parsers.js`, or is the existing coverage we have sufficient?

trentm replied:
Yah, fair, I could definitely improve the test coverage here. Actually before this change there wasn't a single test case checking `error.exception.module`. I'll open a separate ticket to do this in a followup: https://github.com/elastic/apm-agent-nodejs/issues/1918

astorm commented on .tav.yml:
```diff
@@ -242,6 +242,15 @@ koa-router:
   versions: '>=8.0.0'
   commands: node test/instrumentation/modules/elasticsearch.js
+
+# @elastic/elasticsearch
```
👍 on the context for the `versions` choices.

astorm reviewed (approved):
Looks good from here -- previous feedback's been address and it gets us instrumentation for the modern `@elastic/elasticsearch` client. One dealer's choice comment on some additional test coverage, but I think we're good to go as it.  Approving.

delvedor commented on lib/instrumentation/index.js:
```diff
@@ -11,6 +11,7 @@ var shimmer = require('./shimmer')
 var Transaction = require('./transaction')
 
 var MODULES = [
+  '@elastic/elasticsearch',
```
In the coming days I'll publish another module named `@elastic/elasticsearch-canary`, which is identical to `@elastic/elasticsearch`, and it will be used mainly by Kibana.
I'll ping you here as soon as it's available on npm.

trentm commented on .ci/docker/docker-compose.yml:
```diff
@@ -71,7 +71,7 @@ services:
   elasticsearch:
-    image: docker.elastic.co/elasticsearch/elasticsearch:6.6.0
+    image: docker.elastic.co/elasticsearch/elasticsearch:7.10.1
     environment:
```
Yes, that was intentional. I could separate it if we want, but I'd wanted to test this integration against a recent ES.
